File: SummerCal/Services/EventNoteService.swift
import Foundation
import SwiftData

final class EventNoteService {
    func fetchNotes(for eventId: UUID, context: ModelContext) -> [EventNote] {
        let predicate = #Predicate<EventNote> { $0.eventId == eventId }
        let descriptor = FetchDescriptor<EventNote>(predicate: predicate, sortBy: [SortDescriptor(\.createdAt)])
        return (try? context.fetch(descriptor)) ?? []
    }
    
    func addNote(eventId: UUID, body: String, type: EventNoteType = .general, checklistItems: [String] = [], links: [String] = [], context: ModelContext) {
        let note = EventNote(
            eventId: eventId,
            body: body,
            noteType: type,
            checklistItemsJSON: checklistItems.isEmpty ? nil : (try? String(data: JSONEncoder().encode(checklistItems), encoding: .utf8)),
            linksJSON: links.isEmpty ? nil : (try? String(data: JSONEncoder().encode(links), encoding: .utf8))
        )
        context.insert(note)
        try? context.save()
    }
    
    func deleteNote(_ note: EventNote, context: ModelContext) {
        context.delete(note)
        try? context.save()
    }
    
    func updateNote(_ note: EventNote, body: String, context: ModelContext) {
        note.body = body
        note.updatedAt = Date()
        try? context.save()
    }
    
    func addChecklistItem(to note: EventNote, item: String, context: ModelContext) {
        var items = note.checklistItems
        items.append(item)
        note.checklistItemsJSON = (try? String(data: JSONEncoder().encode(items), encoding: .utf8))
        note.updatedAt = Date()
        try? context.save()
    }
    
    func toggleChecklistItem(in note: EventNote, at index: Int, context: ModelContext) {
        var items = note.checklistItems
        guard index < items.count else { return }
        let item = items[index]
        let isChecked = item.hasPrefix("[x] ")
        if isChecked {
            items[index] = String(item.dropFirst(4))
        } else {
            items[index] = "[x] \(item)"
        }
        note.checklistItemsJSON = (try? String(data: JSONEncoder().encode(items), encoding: .utf8))
        note.updatedAt = Date()
        try? context.save()
    }
    
    func addLink(to note: EventNote, link: String, context: ModelContext) {
        var links = note.links
        links.append(link)
        note.linksJSON = (try? String(data: JSONEncoder().encode(links), encoding: .utf8))
        note.updatedAt = Date()
        try? context.save()
    }
    
    func removeLink(from note: EventNote, at index: Int, context: ModelContext) {
        var links = note.links
        guard index < links.count else { return }
        links.remove(at: index)
        note.linksJSON = links.isEmpty ? nil : (try? String(data: JSONEncoder().encode(links), encoding: .utf8))
        note.updatedAt = Date()
        try? context.save()
    }
    
    func notesForNotificationPreview(eventId: UUID, context: ModelContext) -> String {
        let notes = fetchNotes(for: eventId, context: context)
        return notes.prefix(2).map { $0.body }.joined(separator: ". ")
    }
}

File: SummerCalExpo/src/services/EventNoteService.ts
import * as SQLite from 'expo-sqlite';
import { EventNote, EventNoteType } from '../models';
import { v4 as uuidv4 } from 'uuid';

export class EventNoteService {
  async fetchNotesByEventId(db: SQLite.SQLiteDatabase, eventId: string): Promise<EventNote[]> {
    const rows = await db.getAllAsync<any>(
      'SELECT * FROM event_notes WHERE event_id = ? ORDER BY created_at DESC',
      eventId
    );
    return rows.map(row => this.mapNote(row));
  }

  async addNote(
    db: SQLite.SQLiteDatabase,
    eventId: string,
    body: string,
    noteType: EventNoteType = EventNoteType.general,
    checklistItems: string[] = [],
    links: string[] = []
  ): Promise<EventNote> {
    const id = uuidv4();
    const now = new Date().toISOString();
    const checklistJson = JSON.stringify(checklistItems);
    const linksJson = JSON.stringify(links);

    await db.runAsync(
      `INSERT INTO event_notes (id, event_id, body, note_type, checklist_items, links, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
      id, eventId, body, noteType, checklistJson, linksJson, now, now
    );

    return {
      id,
      eventId,
      body,
      noteType,
      checklistItems,
      links,
      createdAt: now,
      updatedAt: now,
    };
  }

  async updateNote(
    db: SQLite.SQLiteDatabase,
    noteId: string,
    updates: { body?: string; noteType?: EventNoteType }
  ): Promise<void> {
    const now = new Date().toISOString();
    const fields: string[] = [];
    const values: any[] = [];

    if (updates.body !== undefined) {
      fields.push('body = ?');
      values.push(updates.body);
    }
    if (updates.noteType !== undefined) {
      fields.push('note_type = ?');
      values.push(updates.noteType);
    }

    if (fields.length === 0) return;

    fields.push('updated_at = ?');
    values.push(now);
    values.push(noteId);

    await db.runAsync(
      `UPDATE event_notes SET ${fields.join(', ')} WHERE id = ?`,
      ...values
    );
  }

  async deleteNote(db: SQLite.SQLiteDatabase, noteId: string): Promise<void> {
    await db.runAsync('DELETE FROM event_notes WHERE id = ?', noteId);
  }

  async addChecklistItem(
    db: SQLite.SQLiteDatabase,
    noteId: string,
    item: string
  ): Promise<string[]> {
    const row = await db.getFirstAsync<any>(
      'SELECT checklist_items FROM event_notes WHERE id = ?',
      noteId
    );
    if (!row) throw new Error(`Note ${noteId} not found`);

    const items: string[] = JSON.parse(row.checklist_items ?? '[]');
    items.push(item);
    const json = JSON.stringify(items);
    const now = new Date().toISOString();

    await db.runAsync(
      'UPDATE event_notes SET checklist_items = ?, updated_at = ? WHERE id = ?',
      json, now, noteId
    );

    return items;
  }

  async toggleChecklistItem(
    db: SQLite.SQLiteDatabase,
    noteId: string,
    index: number
  ): Promise<string[]> {
    const row = await db.getFirstAsync<any>(
      'SELECT checklist_items FROM event_notes WHERE id = ?',
      noteId
    );
    if (!row) throw new Error(`Note ${noteId} not found`);

    const items: string[] = JSON.parse(row.checklist_items ?? '[]');
    if (index >= 0 && index < items.length) {
      const item = items[index];
      if (item.startsWith('[x] ')) {
        items[index] = item.slice(4);
      } else if (item.startsWith('[ ] ')) {
        items[index] = '[x] ' + item.slice(4);
      } else {
        items[index] = '[x] ' + item;
      }
    }
    const json = JSON.stringify(items);
    const now = new Date().toISOString();

    await db.runAsync(
      'UPDATE event_notes SET checklist_items = ?, updated_at = ? WHERE id = ?',
      json, now, noteId
    );

    return items;
  }

  async addLink(
    db: SQLite.SQLiteDatabase,
    noteId: string,
    link: string
  ): Promise<string[]> {
    const row = await db.getFirstAsync<any>(
      'SELECT links FROM event_notes WHERE id = ?',
      noteId
    );
    if (!row) throw new Error(`Note ${noteId} not found`);

    const links: string[] = JSON.parse(row.links ?? '[]');
    links.push(link);
    const json = JSON.stringify(links);
    const now = new Date().toISOString();

    await db.runAsync(
      'UPDATE event_notes SET links = ?, updated_at = ? WHERE id = ?',
      json, now, noteId
    );

    return links;
  }

  async removeLink(
    db: SQLite.SQLiteDatabase,
    noteId: string,
    index: number
  ): Promise<string[]> {
    const row = await db.getFirstAsync<any>(
      'SELECT links FROM event_notes WHERE id = ?',
      noteId
    );
    if (!row) throw new Error(`Note ${noteId} not found`);

    const links: string[] = JSON.parse(row.links ?? '[]');
    if (index >= 0 && index < links.length) {
      links.splice(index, 1);
    }
    const json = JSON.stringify(links);
    const now = new Date().toISOString();

    await db.runAsync(
      'UPDATE event_notes SET links = ?, updated_at = ? WHERE id = ?',
      json, now, noteId
    );

    return links;
  }

  private mapNote(row: any): EventNote {
    return {
      id: row.id,
      eventId: row.event_id,
      body: row.body,
      noteType: (row.note_type as EventNoteType) ?? EventNoteType.general,
      checklistItems: this.parseJsonArray(row.checklist_items),
      links: this.parseJsonArray(row.links),
      createdAt: row.created_at,
      updatedAt: row.updated_at,
    };
  }

  private parseJsonArray(value: string | null | undefined): string[] {
    if (!value) return [];
    try {
      return JSON.parse(value);
    } catch {
      return [];
    }
  }
}
